File: inst/@H5L/H5L.m
## Copyright (C) 2021-2025 Pantxo Diribarne
##
## This program is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program.  If not, see <https://www.gnu.org/licenses/>.

## -*- texinfo -*-
## @deftypefn {} {} H5L
## Collection of pure static methods to manipulate hdf5 files.
## @seealso{}
## @end deftypefn
classdef H5L
  methods
    ## -*- texinfo -*-
    ## @deftypefn {} {} H5L
    ## Dummy class constructor.
    ## @seealso{}
    ## @end deftypefn
    function obj = H5L ()
      error ("H5L: directly call static functions");
    endfunction
  endmethods

  methods (Static = true)

    ## -*- texinfo -*-
    ## @deftypefn {} {[@var{status}, @var{idx_out}, @var{opdata_out}] =} H5L.iterate (@var{group_id}, @var{idx_type}, @var{order}, @var{idx_in},  @var{fcn}, @var{opdata_in})
    ## Iterates over links in a group, with user callback routine, according
    ## to the order within an index.
    ##
    ## @strong{Parameters:}
    ##  @multitable @columnfractions 0.33 0.02 0.65
    ##  @item @var{group_id} @tab @tab Group identifier
    ##  @item @var{idx_type} @tab @tab Index type. One of @{@qcode{"H5_INDEX_NAME"} | @qcode{"H5_INDEX_CRT_ORDER"}@}
    ##  @item @var{order} @tab @tab Iteration order. One of @{@qcode{"H5_ITER_INC"} | @qcode{"H5_ITER_DEC"} | @qcode{"H5_ITER_NATIVE"}@}
    ##  @item @var{idx_in} @tab @tab Iteration index to allow continuing a previous iteration
    ##  @item @var{fcn} @tab @tab Callback function
    ##  @item @var{opdata_in} @tab @tab Callback data
    ##  @end multitable
    ##
    ##
    ## @strong{Return values:}
    ##  @multitable @columnfractions 0.33 0.02 0.65
    ##  @item @var{status} @tab @tab Negative value if the iteration were interrupted
    ##  @item @var{idx_out} @tab @tab Final iteration index
    ##  @item @var{opdata_out} @tab @tab Returned callback data
    ##  @end multitable
    ##
    ## 
    ## @strong{Description:}
    ##
    ## The user callback function signature must be:
    ## @example
    ## @code{[@var{status}, @var{opdata_out}] = fcn (@var{loc_id}, @var{name}, @var{opdata_in})}
    ## @end example
    ## It receives and identifier for the currently visited object @var{loc_id},
    ## its name @var{name} and the data @var{opdata_in} as returned by the
    ## previous iteration.
    ## The function can return a negative @var{status} to instruct
    ## @code{H5L.iterate} to stop iterations. Gathered data are returned in
    ## @var{opdata_out}.
    ##
    ## See original function at 
    ## @url{https://support.hdfgroup.org/documentation/hdf5/latest/group___t_r_a_v.html}.
    ## 
    ## @seealso{H5L.iterate_by_name}
    ## @end deftypefn
    function [status, idx_out, opdata_out] = iterate (varargin)
      [status, idx_out, opdata_out] = __H5L_iterate__ (varargin{:});
    endfunction

    ## -*- texinfo -*-
    ## @deftypefn {} {[@var{status}, @var{idx_out}, @var{opdata_out}] =} H5L.iterateiterate_by_name (@var{group_id}, @var{name}, @var{idx_type}, @var{order}, @var{idx_in},  @var{fcn}, @var{opdata_in}, @var{lapl_id})
    ## Iterates over links in a named group, with user callback routine,
    ## according
    ## to the order within an index.
    ##
    ## @strong{Parameters:}
    ##  @multitable @columnfractions 0.33 0.02 0.65
    ##  @item @var{group_id} @tab @tab Group identifier
    ##  @item @var{name} @tab @tab Group name
    ##  @item @var{idx_type} @tab @tab Index type. One of @{@qcode{"H5_INDEX_NAME"} | @qcode{"H5_INDEX_CRT_ORDER"}@}
    ##  @item @var{order} @tab @tab Iteration order. One of @{@qcode{"H5_ITER_INC"} | @qcode{"H5_ITER_DEC"} | @qcode{"H5_ITER_NATIVE"}@}
    ##  @item @var{idx_in} @tab @tab Iteration index to allow continuing a previous iteration
    ##  @item @var{fcn} @tab @tab Callback function
    ##  @item @var{opdata_in} @tab @tab Callback data
    ##  @item @var{lapl_id} @tab @tab Link access property list identifier.
    ##  @end multitable
    ##
    ##
    ## @strong{Return values:}
    ##  @multitable @columnfractions 0.33 0.02 0.65
    ##  @item @var{status} @tab @tab Negative value if the iteration were interrupted
    ##  @item @var{idx_out} @tab @tab Final iteration index
    ##  @item @var{opdata_out} @tab @tab Returned callback data
    ##  @end multitable
    ##
    ## 
    ## @strong{Description:}
    ##
    ## The user callback function signature must be:
    ## @example
    ## @code{[@var{status}, @var{opdata_out}] = fcn (@var{loc_id}, @var{name}, @var{opdata_in})}
    ## @end example
    ## It receives and identifier for the currently visited object @var{loc_id},
    ## its name @var{name} and the data @var{opdata_in} as returned by the
    ## previous iteration.
    ## The function can return a negative @var{status} to instruct
    ## @code{H5L.iterate} to stop iterations. Gathered data are returned in
    ## @var{opdata_out}.
    ##
    ## See original function at 
    ## @url{https://support.hdfgroup.org/documentation/hdf5/latest/group___t_r_a_v.html}.
    ## 
    ## @seealso{H5L.iterate_by_name}
    ## @end deftypefn
    function [status, idx_out, opdata_out] = iterate_by_name (varargin)
      [status, idx_out, opdata_out] = __H5L_iterate_by_name__ (varargin{:});
    endfunction

  endmethods

endclassdef

File: src/__H5L__.cc
/*

Copyright (C) 2021-2025 Pantxo Diribarne

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

*/

#include <octave/oct.h>
#include <hdf5.h>

#include "./util/h5_oct_util.h"

// PKG_ADD: autoload ("__H5L_iterate__", "__H5L__.oct");
// PKG_DEL: autoload ("__H5L_iterate__", "__H5L__.oct", "remove");
DEFUN_DLD(__H5L_iterate__, args, nargout, 
"-*- texinfo -*-\n\
@deftypefn {} {[@var{status}, @var{idx_out}, @var{opdata_out}] =} H5L.iterate (@var{group_id}, @var{idx_type}, @var{order}, @var{idx_in},  @var{fcn}, @var{opdata_in})\n\
@seealso{}\n\
@end deftypefn")
{
  octave_value_list retval;

  int nargin = args.length ();

  if (nargin != 6)
    print_usage ("H5L.iterate");

  // Group ID
  hid_t group_id = get_h5_id (args, 0, "GROUP_ID", "H5L.iterate");

  // Index type
  H5_index_t idx_type
    = static_cast<H5_index_t> (get_h5_id (args, 1, "IDX_TYPE", "H5L.iterate"));

  // Order
  H5_iter_order_t order
    = static_cast<H5_iter_order_t> (get_h5_id (args, 2, "ORDER",
                                               "H5L.iterate"));

  // Iteration index
  hsize_t idx_in
    = static_cast<hsize_t> (get_h5_id (args, 3, "ORDER", "H5L.iterate"), false);

  // Iteration handler callback
  if (! args(4).is_function_handle ())
    error ("H5L.iterate: FCN must be a function handle");

  user_callback_data ucd;
  ucd.callback = args(4);
  ucd.data = args(5);

  herr_t status = H5Literate (group_id, idx_type, order, &idx_in, iter_handler,
                              &ucd);

  retval.append (octave_value (status));
  retval.append (octave_value (idx_in));
  retval.append (ucd.data);

  return retval;
}

// PKG_ADD: autoload ("__H5L_iterate_by_name__", "__H5L__.oct");
// PKG_DEL: autoload ("__H5L_iterate_by_name__", "__H5L__.oct", "remove");
DEFUN_DLD(__H5L_iterate_by_name__, args, nargout, 
"-*- texinfo -*-\n\
@deftypefn {} {[@var{status}, @var{idx_out}, @var{opdata_out}] =} H5L.iterateiterate_by_name (@var{group_id}, @var{name}, @var{idx_type}, @var{order}, @var{idx_in},  @var{fcn}, @var{opdata_in}, @var{lapl_id})\n\
@seealso{}\n\
@end deftypefn")
{
  octave_value_list retval;

  int nargin = args.length ();

  if (nargin != 8)
    print_usage ("H5L.iterate_by_name");

  // Group ID
  hid_t group_id = get_h5_id (args, 0, "GROUP_ID", "H5L.iterate_by_name");

  // Group name
  std::string name = args(1).xstring_value ("H5L.iterate_by_name: NAME should be a string");

  // Index type
  H5_index_t idx_type
    = static_cast<H5_index_t> (get_h5_id (args, 2, "IDX_TYPE",
                                          "H5L.iterate_by_name"));

  // Order
  H5_iter_order_t order
    = static_cast<H5_iter_order_t> (get_h5_id (args, 3, "ORDER",
                                               "H5L.iterate_by_name"));

  // Iteration index
  hsize_t idx_in
    = static_cast<hsize_t> (get_h5_id (args, 4, "IDX_IN", "H5L.iterate_by_name",
                                       false));

  // Iteration handler callback

  // Pack function handle and user data in a struct and pass it to a generic
  // iteration handler
  if (! args(5).is_function_handle ())
    error ("H5L.iterate_by_name: FCN must be a function handle");

  user_callback_data ucd;
  ucd.callback = args(5);
  ucd.data = args(6);

  // Link access plist
  hsize_t lapl_id = get_h5_id (args, 7, "LAPL_ID", "H5L.iterate_by_name");

  herr_t status = H5Literate_by_name (group_id, name.c_str (), idx_type, order,
                                      &idx_in, iter_handler,
                                      &ucd, lapl_id);

  retval.append (octave_value (status));
  retval.append (octave_value (idx_in));
  retval.append (ucd.data);

  return retval;
}


